Source/Core/Core/Lua/LuaFunctions/LuaRegisters.cpp

```diff
@@ -0,0 +1,504 @@
+#include "Core/Lua/LuaFunctions/LuaRegisters.h"
+
+#include <algorithm>
+#include <fmt/format.h>
+#include <memory>
+#include <string>
+
+#include "Common/CommonTypes.h"
+#include "Core/Lua/LuaHelperClasses/NumberType.h"
+#include "Core/Lua/LuaVersionResolver.h"
+#include "Core/PowerPC/PowerPC.h"
+
+namespace Lua::LuaRegisters
+{
+class LuaRegister
+{
+public:
+  inline LuaRegister() {}
+};
+
+static std::unique_ptr<LuaRegister> lua_register_pointer = nullptr;
+
+LuaRegister* GetLuaRegisterInstance()
+{
+  if (lua_register_pointer == nullptr)
+    lua_register_pointer = std::make_unique<LuaRegister>(LuaRegister());
+  return lua_register_pointer.get();
+}
+
+// Currently supported registers are:
+// r0 - r31, f0 - f31, PC, and LR (the register which stores the return address to jump to when a
+// function call ends)
+void InitLuaRegistersFunctions(lua_State* lua_state, const std::string& lua_api_version)
+{
+  LuaRegister** lua_register_ptr_ptr =
+      (LuaRegister**)lua_newuserdata(lua_state, sizeof(LuaRegister*));
+  *lua_register_ptr_ptr = GetLuaRegisterInstance();
+  luaL_newmetatable(lua_state, "LuaRegistersTable");
+  lua_pushvalue(lua_state, -1);
+  lua_setfield(lua_state, -2, "__index");
+
+  std::array lua_registers_functions_with_versions_attached = {
+      luaL_Reg_With_Version({"getRegister", "1.0", GetRegister}),
+      luaL_Reg_With_Version(
+          {"getRegisterAsUnsignedByteArray", "1.0", GetRegisterAsUnsignedByteArray}),
+      luaL_Reg_With_Version({"getRegisterAsSignedByteArray", "1.0", GetRegisterAsSignedByteArray}),
+      luaL_Reg_With_Version({"setRegister", "1.0", SetRegister}),
+      luaL_Reg_With_Version({"setRegisterFromByteArray", "1.0", SetRegisterFromByteArray})};
+
+  std::unordered_map<std::string, std::string> deprecated_functions_map;
+
+  AddLatestFunctionsForVersion(lua_registers_functions_with_versions_attached, lua_api_version,
+                               deprecated_functions_map, lua_state);
+  lua_setglobal(lua_state, "registers");
+}
+
+class RegisterObject
+{
+public:
+  enum class RegisterType
+  {
+    GeneralPurposeRegister,
+    FloatingPointRegister,
+    PcRegister,
+    ReturnRegister,  // LR register
+    Undefined
+  };
+
+  RegisterObject(RegisterType new_register_type, u8 new_register_number)
+  {
+    register_type = new_register_type;
+    register_number = new_register_number;
+  }
+
+  u8 register_number;
+  RegisterType register_type;
+};
+
+RegisterObject ParseRegister(const char* register_string)
```

Comment:
This function is kind of ugly. Maybe accepting register IDs by string is a bad idea. 

Might be an idea to go with something more strongly typed.

Could do `registers.r(4).read_u32()`, `registers.f(3).write_float(3.1415)`, that way lua gives you a pre-parsed number. 

It wouldn't even be too hard to implement something like `registers.R13.read_u16()`, `registers.PC.write_u32(0x80001000)`. lua's metatables makes it easy to implement something along these lines with just a few functions.

Reply:
I think it looks fine as is, tbh. Plus, it reduces the amount  of documentation, since instead of 68 classes each having 5 methods, we just have 1 class containing 5 methods (which is 68X less functions). This in turn makes the function easier to use.